Terminal.Gui/Application/Application.Mouse.cs

```diff
@@ -56,6 +56,10 @@ public static void UngrabMouse ()
             return;
         }
 
+#if DEBUG_IDISPOSABLE
+        ObjectDisposedException.ThrowIf (MouseGrabView.WasDisposed, typeof (View));
```

Comment:
That view should be the object responsible for throwing this exception, not a class using it.

Same exception is fine - just belongs elsewhere.

If an exception is desired here, it should be most likely be InvalidOperationException, but only if the MouseGrabView View didn't already throw ObjectDisposedException anyway. I'd argue that the need for the exception being thrown by this class is itself a bug, though.

Typically, the object that was disposed is the one who is supposed to throw ODE.

Assuming that principle is followed, and a class like this one that _owns_ a disposable object tries to _use_ that object after that object was disposed, this is how it's supposed to be handled:
 * Highest preference/best option, if possible, is to try to avoid the situation entirely by performing a safe check before trying to use the IDisposable object and creating a new one if necessary. That's one of the main uses of a `Disposed` event - letting subscribers holding a reference to that object know it has been disposed, so they can react appropriately, such as by setting their reference to it to `null`, removing it from any collections they may have put it in, or any other implementation details that the IDisposable doesn't know about and can't be expected to know about; OR
 * Next best is to try to recover from the situation, if you catch an ODE that you weren't able to avoid. Usually that just means making a new instance of the object that was disposed, if there's sufficient information to do so. OR
 * If the object trying to use the IDisposable can't do either of the above safely, it should then do one of:
    * Let the ODE bubble up. OR
    * If it caught the ODE because it has clean-up to do before things can safely continue, re-throw it as-is at the end of the catch. OR
    * If additional context is important to add at that point in the code that the stack trace wouldn't already have been sufficient for, wrap the ODE in an InvalidOperationException, which is the type of exception that is expected to be used when the object is in an invalid state (which it is, _because_ of the disposed object, which is part of its state, and thus the operation is invalid for the current state of the object)
 

All objects owning an IDisposable instance or static member field or property should do the above procedure, at minimum, as part of the contract that IDisposable is supposed to convey.

The IDisposable objects themselves are responsible for their own cleanup, part of which _is_ informing others that disposal has happened.\
That's the easiest, most timely, and most efficient way to do the first bullet (avoiding/preventing the situation entirely), and prevents any exceptions from being thrown, which saves a lot of expensive implicit stuff that happens when an exception is thrown, like generating the stack trace and system event logging and everything else that an exception involves.

At that point, if we've implemented our end of the contract properly, any consumer who doesn't avoid the exception can't blame it on us, as it is not our fault - It's _their_ bug for not respecting the contract, just like it is _currently_ our fault for not respecting our own contract.

Reply:
> That view should be the object responsible for throwing this exception, not a class using it.

My bad. I must have mis-read you before, because I thought you were recommending the opposite.

Reply:
@tig do you want I made any change here?

Reply:
Maybe to this:

`ObjectDisposedException.ThrowIf (MouseGrabView.WasDisposed, MouseGrabView);`

Reply:
> > Maybe to this, because it's passing the object that is failing?
> > `ObjectDisposedException.ThrowIf (MouseGrabView.WasDisposed, MouseGrabView);`
> 
> @dodexahedron is this more appropriate?

Not quite. It's not the exception - it's where it's being thrown.

I can pull this down to take an actual look at it in a few minutes.

Is whatever is latest currently pushed to your branch?

Reply:
> > That view should be the object responsible for throwing this exception, not a class using it.
> 
> My bad. I must have mis-read you before, because I thought you were recommending the opposite.

I was pretty certain there would be at least one or five reversals. 😅

That's part of why those comments took me a while to write - I had to keep going back and making sure I wasn't reversing things, which was super easy to do and I _did_ do (but therefore fixed before I posted....I think). 😆 

The pseudo-logic way I stuck it in this thread is a lot easier for _my_ eye to follow, at least, than how I babbled it out in your PR. 🤔